h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -2152,45 +2175,70 @@ public void computeImpl() {
-      // lambda search loop
-      for (int i = 0; i < _parms._lambda.length; ++i) {  // lambda search
-        if (_job.stop_requested() || (timeout() && _model._output._submodels.length > 0) || _earlyStop) break;  //need at least one submodel on timeout to avoid issues.
-        if(_parms._max_iterations != -1 && _state._iter >= _parms._max_iterations) break;
-        Submodel sm = computeSubmodel(i,_parms._lambda[i], nullDevTrain, nullDevTest);
-        double trainDev = sm.devianceTrain;
-        double testDev = sm.devianceTest;
-        devHistoryTest[i % devHistoryTest.length] = (oldDevTest - testDev)/oldDevTest;
-        oldDevTest = testDev;
-        devHistoryTrain[i % devHistoryTrain.length] = (oldDevTrain - trainDev)/oldDevTrain;
-        oldDevTrain = trainDev;
-        if(_parms._lambda[i] < _lmax && Double.isNaN(_lambdaCVEstimate) && !_earlyStopEnabled /** if we have cv lambda estimate we should use it, can not stop before reaching it */) {
-          if (_parms._early_stopping && _state._iter >= devHistoryTrain.length) {
-            double s = ArrayUtils.maxValue(devHistoryTrain);
-            if (s < 1e-4) {
-              Log.info(LogMsg("converged at lambda[" + i + "] = " + _parms._lambda[i] + ", improvement on train = " + s));
-              break; // started overfitting
-            }
-            if (_validDinfo != null && _parms._nfolds <= 1) { // check for early stopping on test but only if not doing xval
-              s = ArrayUtils.maxValue(devHistoryTest);
-              if (s < 0) {
-                Log.info(LogMsg("converged at lambda[" + i + "] = " + _parms._lambda[i] + ", improvement on test = " + s));
+      if (!_parms._HGLM) {  // only need these for non HGLM
+        _ginfoStart = GLMUtils.copyGInfo(_state.ginfo());
+        _betaDiffStart = _state.getBetaDiff();
+      }
+      // alpha, lambda search loop
+      int submodelCount = 0;
+      _model._output._lambda_array_size = _parms._lambda.length;
+      for (int alphaInd = 0; alphaInd < _parms._alpha.length; alphaInd++) {
+        _state.setAlpha(_parms._alpha[alphaInd]);   // loop through the alphas
+        if ((!_parms._HGLM) && (alphaInd > 0)) // no need for cold start during the first iteration
+          coldStart(devHistoryTrain, devHistoryTest);  // reset beta, lambda, currGram
+        for (int i = 0; i < _parms._lambda.length; ++i) {  // for lambda search, can quit before it is done
+          if (_job.stop_requested() || (timeout() && _model._output._submodels.length > 0))
+            break;  //need at least one submodel on timeout to avoid issues.
+          if (_parms._max_iterations != -1 && _state._iter >= _parms._max_iterations) 
+            break;// iterations accumulate across all lambda/alpha values
+          if ((!_parms._HGLM && (_parms._cold_start || (!_parms._lambda_search && _parms._cold_start))) && (i > 0)) // default: cold_start for non lambda_search
+            coldStart(devHistoryTrain, devHistoryTest);
```

Comment:
coldStart will reset _state._iter to 0, is that correct? how are we counting the number of iterations if it is reset here?

Reply:
The meaning of cold start is to treat each parameter combination of alpha and lambda as a new model that has nothing to do with the previous model being built.  Hence, apart from reseting the coefficients, gram matrix and others, we also need to reset the model.iteration number.  Otherwise, we are counting the iteration number as a cumulation of all the previous model's iteration counts which I think is incorrect.  However, during lambda_search or when cold_start is not set, the model iteration number will accumulate across all models built.

Reply:
@michalk:  When you give me an array of alpha and lambda values and cold_start = True, I assume you want me to return a number of GLM models that are built independent of the other GLM models.  You can still control the max_iterations such that each model being built will not exceed that max_iterations.  As far as I am concerned, when you set cold-start = true, you are looking for independent models built and returned to you.  Does this make sense to you?